sql/core/src/main/scala/org/apache/spark/sql/execution/r/MapPartitionsRWrapper.scala

```diff
@@ -25,6 +25,21 @@ import org.apache.spark.sql.Row
 import org.apache.spark.sql.types.{BinaryType, StructField, StructType}
 
 /**
+ * A function wrapper that applies the given R function to each partition of each group.
+ */
+private[sql] case class MapGroupsRWrapper(
```

Comment:
Maybe can do some optimization to above idea. As it needs a buffer and reduces the parallelism. Can improve R worker, so that multiple groups can be passed to R worker one by one ( this way, R worker knows the group boundary),  and In the R worker, the R gapply function will be call multiple times, one for each group. Results are flattened and merged inside the R worker.